internal/sidekick/internal/parser/discovery/discovery.go

```diff
@@ -0,0 +1,473 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package discovery
+
+import (
+	"encoding/json"
+	"fmt"
+	"maps"
+	"slices"
+	"sort"
+	"strings"
+
+	"github.com/googleapis/librarian/internal/sidekick/internal/api"
+	"github.com/googleapis/librarian/internal/sidekick/internal/parser/svcconfig"
+	"google.golang.org/genproto/googleapis/api/serviceconfig"
+)
+
+// NewAPI parses the discovery doc in `contents` and returns the corresponding `api.API` model.
+func NewAPI(serviceConfig *serviceconfig.Service, contents []byte) (*api.API, error) {
+	doc, err := newDiscoDocument(contents)
+	if err != nil {
+		return nil, err
+	}
+	result := &api.API{
+		Name:        doc.Name,
+		Title:       doc.Title,
+		Description: doc.Description,
+		Messages:    make([]*api.Message, 0),
+		State: &api.APIState{
+			ServiceByID: make(map[string]*api.Service),
+			MethodByID:  make(map[string]*api.Method),
+			MessageByID: make(map[string]*api.Message),
+			EnumByID:    make(map[string]*api.Enum),
+		},
+	}
+
+	// Discovery docs do not define a service name or package name. The service
+	// config may provide one.
+	serviceName := "Service"
```

Comment:
I remember in C++ we mapped each resource to a service.

Thinking ahead, we feature-gate the services in large crates like `aiplatform`. I think compute would need something similar. We *could* feature-gate the APIs on a single compute client, but I am skeptical.

Reply:
I agree, I removed the placeholder.